contracts/src/batch-bridge/L2BatchBridgeGateway.sol

```diff
@@ -0,0 +1,239 @@
+     *************/
+
+    modifier onlyMessenger() {
+        if (_msgSender() != messenger) {
+            revert ErrorCallerNotMessenger();
+        }
+        _;
+    }
+
+    /***************
+     * Constructor *
+     ***************/
+
+    /// @param _counterpart The address of `L1BatchBridgeGateway` contract in L1.
+    /// @param _messenger The address of `L2ScrollMessenger` contract in L2.
+    constructor(address _counterpart, address _messenger) {
+        _disableInitializers();
+
+        counterpart = _counterpart;
+        messenger = _messenger;
+    }
+
+    /// @notice Initialize the storage of `L2BatchBridgeGateway`.
+    function initialize() external initializer {
+        __Context_init(); // from ContextUpgradeable
+        __ERC165_init(); // from ERC165Upgradeable
+        __AccessControl_init(); // from AccessControlUpgradeable
+        __AccessControlEnumerable_init(); // from AccessControlEnumerableUpgradeable
+
+        _grantRole(DEFAULT_ADMIN_ROLE, _msgSender());
+    }
+
+    /*****************************
+     * Public Mutating Functions *
+     *****************************/
+
+    /// @notice Receive batch bridged ETH from `L2ScrollMessenger`.
+    receive() external payable onlyMessenger {
+        // empty
+    }
+
+    /// @notice Finalize L1 initiated batch token bridge.
+    /// @param l1Token The address of the token in L1.
+    /// @param l2Token The address of the token in L2.
+    /// @param phase The phase of this batch bridge.
+    /// @param hash The hash of this phase.
+    function finalizeBatchBridge(
+        address l1Token,
+        address l2Token,
+        uint256 phase,
+        bytes32 hash
+    ) external onlyMessenger {
+        // this usually won't happen, check just in case.
+        if (tokenMapping[l2Token] != l1Token) {
+            revert ErrorL1TokenMismatched();
+        }
+        if (counterpart != IL2ScrollMessenger(messenger).xDomainMessageSender()) {
+            revert ErrorMessageSenderNotCounterpart();
+        }
+
+        phaseHashes[l2Token][phase] = hash;
+
+        emit FinalizeBatchBridge(l1Token, l2Token, phase);
+    }
+
+    /************************
+     * Restricted Functions *
+     ************************/
+
+    /// @notice Update layer 2 to layer 1 token mapping.
+    /// @param l2Token The address of corresponding ERC20 token on layer 2.
+    /// @param l1Token The address of ERC20 token on layer 1.
+    function updateTokenMapping(address l2Token, address l1Token) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        address oldL1Token = tokenMapping[l2Token];
+        tokenMapping[l2Token] = l1Token;
+
+        emit UpdateTokenMapping(l2Token, oldL1Token, l1Token);
+    }
+
+    /// @notice Withdraw distribution failed tokens.
+    /// @param token The address of token to withdraw.
+    /// @param receiver The address of token receiver.
+    function withdrawFailedAmount(address token, address receiver) external onlyRole(DEFAULT_ADMIN_ROLE) {
```

Comment:
if there is an event for this, will be much easier to track the withdrawal status of failed balance (clear the sum added based on `DistributeFailed` event).